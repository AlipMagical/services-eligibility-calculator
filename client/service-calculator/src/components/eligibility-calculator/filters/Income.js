/**
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Col, Row } from 'react-bootstrap';
import React, { useState } from 'react';

import { Chip } from '../Chip';
import FilterModal from './FilterModal';
import PropTypes from 'prop-types';
import ReactGA from 'react-ga';
import Slider from '@material-ui/core/Slider';
import ValueLabel from '@material-ui/core/Slider/ValueLabel';
import { useTranslation } from 'react-i18next';

const INCOME_MAX = 99;

const IncomeSliderLabel = props => {
  let label = props.value.toString();
  if (props.value > 0) {
    label += 'K';
  }
  if (props.value === INCOME_MAX) {
    label += '+';
  }
  return <ValueLabel {...props} value={label} />;
};

const IncomeChipText = ({ income }) => {
  const { t } = useTranslation();
  if (income === undefined) {
    return t('filterIncome.buttonAdd');
  }
  let label = `$${income}`;
  if (income > 0) {
    label += ',000';
  }
  if (income === INCOME_MAX) {
    label += '+';
  }
  return <>{label}</>;
};

function sendAnalytics(income) {
  ReactGA.event({
    category: 'Filter',
    action: 'Income',
    value: income * 1000
  });
}

function IncomeFilterModal({ income, setIncome, setShow }) {
  const [sliderState, setSliderState] = useState(income);
  const { t } = useTranslation();

  return (
    <FilterModal
      className="income"
      show
      onApply={() => {
        setIncome(sliderState);
        setShow(false);
        sendAnalytics(sliderState);
      }}
      onHide={() => setShow(false)}
      onClear={() => {
        setShow(false);
        setIncome(undefined);
      }}
      clearDisabled={sliderState === undefined}
      title={t('filterIncome.modalTitle')}
    >
      <Row>
        <Col xs={12} className="mt-1 mb-4">
          <span className="input-label">
            {t('filterIncome.modalDescription')}
          </span>
        </Col>
        <Col xs={12} className="income-slider-col">
          <Slider
            ValueLabelComponent={IncomeSliderLabel}
            min={0}
            max={99}
            step={1}
            value={sliderState || 0}
            valueLabelDisplay="on"
            onChange={(_, value) => {
              setSliderState(value);
            }}
            aria-label={t('filterIncome.label')}
            getAriaValueText={value =>
              value +
              (value > 0 && ' ' + t('filterIncome.thousand')) +
              ((value === INCOME_MAX && '+') || '') +
              ' ' +
              t('filterIncome.dollars')
            }
          ></Slider>
        </Col>
      </Row>
    </FilterModal>
  );
}

export default function Income({ income, setIncome }) {
  const [show, setShow] = useState(false);

  return (
    <>
      <Chip
        analytics={{ action: 'Income', label: 'Add income' }}
        active={income !== undefined}
        onClick={() => setShow(true)}
      >
        <IncomeChipText income={income} />
      </Chip>
      {show && <IncomeFilterModal {...{ income, setIncome, setShow }} />}
    </>
  );
}

Income.propTypes = {
  income: PropTypes.number,
  setIncome: PropTypes.func.isRequired
};
