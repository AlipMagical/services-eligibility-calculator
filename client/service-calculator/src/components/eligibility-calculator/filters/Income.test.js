/**
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import { act } from 'react-dom/test-utils';
import Income from './Income';
import { axe, toHaveNoViolations } from 'jest-axe';
import { render, fireEvent, waitForElement } from '@testing-library/react';

expect.extend(toHaveNoViolations);

it('renders button without selected income', async () =>
  act(async () => {
    const { container } = render(
      <Income income={undefined} setIncome={() => {}} />
    );

    expect(container).toMatchSnapshot();
    expect(await axe(container)).toHaveNoViolations();
  }));

it('renders modal without selected income', async () =>
  act(async () => {
    const { getByText, getByRole } = render(
      <Income income={undefined} setIncome={() => {}} />
    );

    fireEvent.click(getByText('Add income'));

    const modal = await waitForElement(() => getByRole('document'));

    expect(modal).toMatchSnapshot();
    expect(await axe(modal)).toHaveNoViolations();
  }));

it('renders button with selected income', async () =>
  act(async () => {
    const { container } = render(<Income income={0} setIncome={() => {}} />);

    expect(container).toMatchSnapshot();
    expect(await axe(container)).toHaveNoViolations();
  }));

it('renders modal with selected income', async () =>
  act(async () => {
    const { getByText, getByRole } = render(
      <Income income={99} setIncome={() => {}} />
    );

    fireEvent.click(getByText('$99,000+'));

    const modal = await waitForElement(() => getByRole('document'));

    expect(modal).toMatchSnapshot();
    expect(await axe(modal)).toHaveNoViolations();
  }));
