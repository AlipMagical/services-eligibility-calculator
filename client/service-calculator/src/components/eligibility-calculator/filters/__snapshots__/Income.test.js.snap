// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders button with selected income 1`] = `
<div>
  <button
    class="btn btn-chip ellipsis active"
  >
    $0
  </button>
</div>
`;

exports[`renders button without selected income 1`] = `
<div>
  <button
    class="btn btn-chip ellipsis "
  >
    Add income
  </button>
</div>
`;

exports[`renders modal with selected income 1`] = `
<div
  class="modal-dialog modal-md modal-dialog-centered"
  role="document"
>
  <div
    class="modal-content"
  >
    <div
      class="modal-header"
    >
      <div
        class="text-center w-100 text-center modal-title h4"
      >
        Income
        <button
          aria-label="Close"
          class="close float-right modal-close"
          data-dismiss="modal"
          type="button"
        >
          <span
            aria-hidden="true"
            class="align-middle"
          >
            ×
          </span>
        </button>
      </div>
    </div>
    <div
      class="modal-body"
    >
      <div
        class="row"
      >
        <div
          class="mt-1 mb-4 col-12"
        >
          <span
            class="input-label"
          >
            Include yearly income from everyone in your household
          </span>
        </div>
        <div
          class="income-slider-col col-12"
        >
          <span
            class="MuiSlider-root MuiSlider-colorPrimary"
          >
            <span
              class="MuiSlider-rail"
            />
            <span
              class="MuiSlider-track"
              style="left: 0%; width: 100%;"
            />
            <input
              type="hidden"
              value="99"
            />
            <span
              aria-label="Annual household income"
              aria-orientation="horizontal"
              aria-valuemax="99"
              aria-valuemin="0"
              aria-valuenow="99"
              aria-valuetext="99 thousand+ dollars"
              class="MuiSlider-thumb MuiSlider-thumbColorPrimary PrivateValueLabel-open-47 PrivateValueLabel-thumb-46"
              data-index="0"
              role="slider"
              style="left: 100%;"
              tabindex="0"
            >
              <span
                class="PrivateValueLabel-offset-48 MuiSlider-valueLabel"
              >
                <span
                  class="PrivateValueLabel-circle-49"
                >
                  <span
                    class="PrivateValueLabel-label-50"
                  >
                    99K+
                  </span>
                </span>
              </span>
            </span>
          </span>
        </div>
      </div>
    </div>
    <div
      class="modal-footer"
    >
      <button
        class="btn btn-clear btn-gutter my-0"
      >
        Clear
      </button>
      <button
        class="btn btn-primary float-right my-0 mx-0"
      >
        Apply
      </button>
    </div>
  </div>
</div>
`;

exports[`renders modal without selected income 1`] = `
<div
  class="modal-dialog modal-md modal-dialog-centered"
  role="document"
>
  <div
    class="modal-content"
  >
    <div
      class="modal-header"
    >
      <div
        class="text-center w-100 text-center modal-title h4"
      >
        Income
        <button
          aria-label="Close"
          class="close float-right modal-close"
          data-dismiss="modal"
          type="button"
        >
          <span
            aria-hidden="true"
            class="align-middle"
          >
            ×
          </span>
        </button>
      </div>
    </div>
    <div
      class="modal-body"
    >
      <div
        class="row"
      >
        <div
          class="mt-1 mb-4 col-12"
        >
          <span
            class="input-label"
          >
            Include yearly income from everyone in your household
          </span>
        </div>
        <div
          class="income-slider-col col-12"
        >
          <span
            class="MuiSlider-root MuiSlider-colorPrimary"
          >
            <span
              class="MuiSlider-rail"
            />
            <span
              class="MuiSlider-track"
              style="left: 0%; width: 0%;"
            />
            <input
              type="hidden"
              value="0"
            />
            <span
              aria-label="Annual household income"
              aria-orientation="horizontal"
              aria-valuemax="99"
              aria-valuemin="0"
              aria-valuenow="0"
              aria-valuetext="0 dollars"
              class="MuiSlider-thumb MuiSlider-thumbColorPrimary PrivateValueLabel-open-22 PrivateValueLabel-thumb-21"
              data-index="0"
              role="slider"
              style="left: 0%;"
              tabindex="0"
            >
              <span
                class="PrivateValueLabel-offset-23 MuiSlider-valueLabel"
              >
                <span
                  class="PrivateValueLabel-circle-24"
                >
                  <span
                    class="PrivateValueLabel-label-25"
                  >
                    0
                  </span>
                </span>
              </span>
            </span>
          </span>
        </div>
      </div>
    </div>
    <div
      class="modal-footer"
    >
      <button
        class="disabled btn btn-clear btn-gutter my-0"
      >
        Clear
      </button>
      <button
        class="btn btn-primary float-right my-0 mx-0"
      >
        Apply
      </button>
    </div>
  </div>
</div>
`;
